src/dp_maglev.c

```diff
@@ -0,0 +1,210 @@
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and IronCore contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <arpa/inet.h>
+#include "dp_error.h"
+#include "rte_rib6.h"
+#include "dp_maglev.h"
+
+#include <stdint.h>
+
+#define DP_MURMURHASH2_MAGIC	0x5bd1e995
+#define DP_DJB_HASH_MAGIC		5381
+#define DP_MAGLEV_POS_FREE		-1
+#define DP_SHIFT_UP				true
+#define DP_SHIFT_DOWN			false
+
+static uint32_t dp_murmur_hash2(const uint8_t ipv6[])
+{
+	uint32_t len = DP_IPV6_ADDR_SIZE;
+	const uint8_t *data = ipv6;
+	uint32_t h = 0 ^ len;
+	uint32_t k;
+
+	while (len >= 4) {
+		k = data[0] | (data[1] << 8) | (data[2] << 16) | (data[3] << 24);
+
+		k *= DP_MURMURHASH2_MAGIC;
+		k ^= k >> 24;
+		k *= DP_MURMURHASH2_MAGIC;
+
+		h *= DP_MURMURHASH2_MAGIC;
+		h ^= k;
+
+		data += 4;
+		len -= 4;
+	}
+
+	switch (len) {
+	case 3:
+		h ^= data[2] << 16; // Fall through
+	case 2:
+		h ^= data[1] << 8;  // Fall through
+	case 1:
+		h ^= data[0];
+		h *= DP_MURMURHASH2_MAGIC;
+		break;
+	default:
+		break;
+	}
+
+	h ^= h >> 13;
+	h *= DP_MURMURHASH2_MAGIC;
+	h ^= h >> 15;
+
+	return h;
+}
+
+static uint32_t dp_djb_hash(const uint8_t ipv6[])
+{
+	uint32_t hash = DP_DJB_HASH_MAGIC;
+	const uint8_t *byte = ipv6;
+	int c;
+
+	while ((c = *byte++))
+		hash = ((hash << 5) + hash) + c;
+
+	hash &= ~(1 << 31);
+	return hash;
+}
+
+static int *dp_maglev_permutation(struct lb_value *lbval)
+{
+	int  i, j, size, *permutation;
+	uint32_t offset, skip;
+
+	size = lbval->back_end_cnt * DP_LB_MAGLEV_LOOKUP_SIZE;
+	permutation = rte_zmalloc("maglev_perm", sizeof(int) * size, RTE_CACHE_LINE_SIZE);
```

Comment:
Both here and the other allocation below do not log anything, though improbable, it would still be helpful to have in the log.